File: 437-path-sum-iii/path-sum-iii.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function pathSum(root: TreeNode | null, targetSum: number): number {
    const prefixSums: Map<number, number> = new Map()
    prefixSums.set(0, 1)
    let count: number = 0

    function dfs(node: TreeNode | null, currentSum: number): void {
        if (root === null) return

        currentSum += node.val
        const neededSum: number = currentSum - targetSum
        if (prefixSums.has(neededSum)) {
            count += prefixSums.get(neededSum)
        }

        prefixSums.set(currentSum, (prefixSums.get(currentSum) || 0) + 1)

        if (node.left !== null) dfs(node.left, currentSum)
        if (node.right !== null) dfs(node.right, currentSum)

        prefixSums.set(currentSum, prefixSums.get(currentSum) - 1)
    }

    dfs(root, 0)
    return count
};

File: 437-path-sum-iii/path-sum-iii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> int:
        if not root:
            return 0

        self.count = 0

        prefix_sum = {0: 1}

        self.dfs(root, 0, targetSum, prefix_sum)

        return self.count

    def dfs(self, node, current_sum, target, prefix_sum):
        if not node:
            return

        current_sum += node.val

        self.count += prefix_sum.get(current_sum - target, 0)

        prefix_sum[current_sum] = prefix_sum.get(current_sum, 0) + 1

        self.dfs(node.left, current_sum, target, prefix_sum)
        self.dfs(node.right, current_sum, target, prefix_sum)

        prefix_sum[current_sum] -= 1
        if prefix_sum[current_sum] == 0:
            del prefix_sum[current_sum]